
Task 1: Data Cleaning & Preprocessing

Dataset: Titanic Dataset

Objective:  
Clean and prepare raw data for analysis

Tasks:
1. Handle missing values
2. Remove duplicates
3. Convert data types
4. Rename columns

Output:
       Clean dataset ready for analysis



Step 1. Load Dataset


In [2]:
import pandas as pd
import numpy as np
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df_titanic = pd.read_csv(url)
df_titanic.head() #show the first 5 rows of the dataset.

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


STEP 2: Understand Data

In [3]:
df_titanic.shape #(Check the number of rows and columns in the dataset , 891 rows and 12 columns)

(891, 12)

In [4]:
df_titanic.columns   # to view the name of the columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [5]:
df_titanic.info()  #info about the dataset and its datatype.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [6]:
df_titanic.isnull().sum()   #Check the null values or missing values in the dataset

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


In [7]:
df_titanic.duplicated().sum()   # check number of duplicates. duplicates means - A row that is exactly the same as a previous row

np.int64(0)

STEP 3: Handle Missing Values

What is Missing Value Handling?

👉 In real datasets, some values are empty (NaN)

👉 ML models cannot work with missing values

So we must:

Fill (Impute)
OR
Drop

Here the missing values are in the following columns.

1. Age - 177

2. cabin - 687

3. Embarked - 2.

Why Age has missing values?

    Some passengers’ age not recorded.

 Why NOT drop rows?

If you drop:

1. You may lose too much data.

2. Dataset becomes smaller.

Why use Median?

Ex:

Age values:

[22, 25, 30, 35, 100]

Mean = 42.4 ❌ (affected by outlier 100)

Median = 30 ✅ (middle value)


✅ Final Reason

👉 Median is robust to outliers

In [8]:
# to fill the missing values of Age.
df_titanic['Age'] = df_titanic['Age'].fillna(df_titanic['Age'].median())
df_titanic.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Now: Handle Cabin Column.

Step 1: Check missing values


687 missing out of 891
🧠 Why Cabin is a problem?

👉 Around 77% data is missing

👉 That means:

Very little useful information

Filling will be inaccurate

🚫 Why NOT fill Cabin?

❌ Filling with mode/median?

Cabin is not numeric

Values are like: C85, E46 (not meaningful to average)

❌ Filling with random value?

Completely misleading

✅ Best Solution → DROP COLUMN

In [9]:
df_titanic.drop('Cabin', axis=1, inplace=True)  # axis = 1 means column and inplace = True means modify original frame.
df_titanic.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Embarked'],
      dtype='object')

Handle Embarked.

What is Embarked?

👉 It represents:

Port where passenger boarded
Values:
C → Cherbourg
Q → Queenstown
S → Southampton

👉 This is a categorical column

--->Only 2 missing values.

How to handle missing values?

✅ Use MODE (most frequent value)

Why mode?

Because:

Data is categorical

Cannot use mean or median

In [10]:
df_titanic['Embarked'] = df_titanic['Embarked'].fillna(df_titanic['Embarked'].mode()[0])

In [11]:
#after handling the missing value, we get,
df_titanic.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,0
SibSp,0
Parch,0
Ticket,0
Fare,0


In [12]:
df_titanic.duplicated().sum()

np.int64(0)

step 4 convert data types


In [13]:
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


3. Pclass → int64 ❗

👉 Needs change:

Why important?

Even though values are numbers:

1, 2, 3 are NOT quantities

They are categories (class types)

👉 So:

❌ Wrong → treat as numeric

✅ Correct → treat as categorical

In [16]:
df_titanic['Pclass'] = df_titanic['Pclass'].astype('category')
df_titanic['Sex'] = df_titanic['Sex'].astype('category')
df_titanic['Embarked'] = df_titanic['Embarked'].astype('category')

In [17]:
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    int64   
 1   Survived     891 non-null    int64   
 2   Pclass       891 non-null    category
 3   Name         891 non-null    object  
 4   Sex          891 non-null    category
 5   Age          891 non-null    float64 
 6   SibSp        891 non-null    int64   
 7   Parch        891 non-null    int64   
 8   Ticket       891 non-null    object  
 9   Fare         891 non-null    float64 
 10  Embarked     891 non-null    category
dtypes: category(3), float64(2), int64(4), object(2)
memory usage: 58.8+ KB


Step5 : Rename Columns

👉 Why rename?
Improve readability

Follow naming standards

Make code professional

🧠 Current column names

PassengerId, Survived, Pclass, SibSp, Parch

👉 Not very clean / consistent

🎯 Best Practice Naming

Use:

lowercase

underscore _

meaningful names

In [19]:
df_titanic.rename(columns={
    'PassengerId': 'passenger_id',
    'Survived' : 'survival_status',
    'Pclass': 'passenger_class',
    'SibSp': 'sibling_spouse',
    'Parch': 'parent_child'

}, inplace = True)
df_titanic.columns

Index(['passenger_id', 'survival_status', 'passenger_class', 'Name', 'Sex',
       'Age', 'sibling_spouse', 'parent_child', 'Ticket', 'Fare', 'Embarked'],
      dtype='object')

# step 6.Save cleaned dataset

In [20]:
df_titanic.to_csv('cleaned_titanic.csv', index=False)
from google.colab import files
files.download('cleaned_titanic.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>